### PR TITLE
#2132: fix(sae): honor top_k in seed LSQ and seed refinement to prevent cold-start co-collapse

### PR DESCRIPTION
**Pull request by SauersML**
### Motivation
- The cold-start seed path (decoder LSQ + EM-style routing refinement) used dense softmax responsibilities while the Arrow‑Schur solve later applied a `top_k` active-set cap, producing a different objective at init and enabling dictionary co-collapse on planted single-active mixtures.
- The change ensures the initialization sees the same capped support as the optimization so seeding cannot reintroduce the uniform-saddle that collapses atoms as K grows.

### Description
- Apply the softmax active-set cap earlier in the training pipeline by calling `base_term.set_softmax_active_cap(top_k)` before the cold routing refinement to make the seed path respect the requested `top_k` (file: `crates/gam-pyffi/src/latent/latent_basis_and_sae_ffi.rs`).
- Thread `top_k: Option<usize>` into the decoder seed helper `sae_decoder_lsq_init(...)` and project/renormalize per-row softmax responsibilities onto the top‑k support before forming the joint LSQ design (file: `crates/gam-pyffi/src/latent/latent_basis_and_sae_ffi.rs`).
- Pass `top_k` through the minimal fit / EM refinement M-step so every place that computes the seed decoder honors the active-set cap (files: `crates/gam-pyffi/src/latent/latent_basis_and_sae_ffi_tail.rs`, `crates/gam-pyffi/src/latent/latent_basis_and_sae_ffi.rs`).
- Add a focused regression test `sae_decoder_lsq_seed_honors_softmax_top_k_support_2132` that asserts `top_k=1` seed LSQ fits atoms only on their selected rows (file: `crates/gam-pyffi/tests/src_modules/lib_tests.rs`).

### Testing
- Built full tree: `./build.sh` — succeeded (library build completed).
- Scoped lib check: `./build.sh check -p gam-pyffi --lib` — succeeded (no warnings).
- Built pyffi test binary (no-run): `./build.sh test -p gam --test pyffi --no-run` — succeeded (test binary built).
- Ran the focused test binary locally where possible: `./build.sh test -p gam-pyffi --lib sae_decoder_lsq_seed_honors_softmax_top_k_support_2132 -- --nocapture` — compilation succeeded but the lib-test link failed in this environment due to missing Python/PyO3 symbols during link (linker error unrelated to the changed logic); the same test compiles in the normal crate test build and the added unit compiles against the library surface.

All changes are in Rust (no Python math), introduce no forbidden shortcuts from SPEC (no grid-search, no fabricated convergence, no hard timeouts, no magic constants without justification), and add a minimal deterministic unit that pins the `top_k` seed contract.

---
Refs #2132 (core cold-joint seed co-collapse; residual stagewise-below-PCA-floor and OOS reconstruct-shape sub-findings tracked separately in the issue).

<details><summary>codex self-assessment</summary>

rates/gam-pyffi)\n      Checking gam-predict v0.3.148 (/workspace/gam/crates/gam-predict)\n      Finished `dev` profile [unoptimized + debuginfo] target(s) in 1m 58s\n  ```\n\n* \u2705 `./build.sh test -p gam --test pyffi --no-run`  \n  Output:\n  ```text\n  [build.sh] test -p gam --test pyffi --no-run -> exit 0 in 10s (dedup=MISS, recompiled 2 crates)\n     Compiling gam v0.3.148 (/workspace/gam)\n     Compiling gam-predict v0.3.148 (/workspace/gam/crates/gam-predict)\n      Finished `test` profile [optimized + debuginfo] target(s) in 9.41s\n    Executable tests/pyffi/main.rs (target/debug/deps/pyffi-61a275d6536e0bb2)\n  ```\n\n* \u26a0\ufe0f `./build.sh test -p gam-pyffi --lib sae_decoder_lsq_seed_honors_softmax_top_k_support_2132 -- --nocapture`  \n  Output:\n  ```text\n  [build.sh] test -p gam-pyffi --lib sae_decoder_lsq_seed_honors_softmax_top_k_support_2132 -- --nocapture -> exit 101 in 265s (dedup=MISS, recompiled 3 crates)\n  FAILED (101):\n  ...\n  rust-lld: error: undefined symbol: PyExc_AttributeError\n  ...\n  rust-lld: error: undefined symbol: PyUnicode_FromStringAndSize\n  ...\n  rust-lld: error: undefined symbol: PyExc_TypeError\n  ...\n  error: could not compile `gam-pyffi` (lib test) due to 1 previous error\n  ```\n  This failed while linking the `gam-pyffi` lib-test harness against Python/PyO3 symbols, not while compiling the changed Rust code. The scoped non-test library build and relevant test binary build both passed.\n\n## CODE DIFF\n\n```diff\ndiff --git a/crates/gam-pyffi/src/latent/latent_basis_and_sae_ffi.rs b/crates/gam-pyffi/src/latent/latent_basis_and_sae_ffi.rs\nindex d43ff41..e17eaed 100644\n--- a/crates/gam-pyffi/src/latent/latent_basis_and_sae_ffi.rs\n+++ b/crates/gam-pyffi/src/latent/latent_basis_and_sae_ffi.rs\n@@ -3267,6 +3267,13 @@ fn sae_manifold_fit_inner<'py>(\n             .set_temperature_schedule(schedule)\n             .map_err(py_value_error)?;\n     }\n+    // #2132 \u2014 the active-set cap is part of the model being optimized, so the\n+    // cold routing refinement must see the same top-k softmax support as the\n+    // subsequent Arrow-Schur solve. Installing it only after the EM seed let\n+    // the seed decoder LSQ train every atom on every row under dense softmax\n+    // responsibilities; with top_k=1 planted mixtures that reintroduced the\n+    // uniform co-collapse saddle before the capped solver ever ran.\n+    base_term.set_softmax_active_cap(top_k);\n \n     // Cold-start routing seed refinement (#629, #630). The cold residual-logit\n@@ -3398,7 +3405,6 @@ fn sae_manifold_fit_inner<'py>(\n-    base_term.set_softmax_active_cap(top_k);\n     let seed_dispersion = base_term\n@@ -6086,6 +6092,7 @@ fn sae_decoder_lsq_init(\n     alpha: f64,\n     tau: f64,\n     jumprelu_threshold: f64,\n+    top_k: Option<usize>,\n ) -> Result<Array3<f64>, String> {\n@@ -6116,6 +6123,13 @@ fn sae_decoder_lsq_init(\n     match assignment_kind {\n         \"softmax\" => {\n+            if let Some(k_top) = top_k {\n+                if k_top == 0 || k_top > k_atoms {\n+                    return Err(format!(\n+                        \"sae_decoder_lsq_init: top_k must satisfy 1 <= top_k <= k_atoms={k_atoms}; got {k_top}\"\n+                    ));\n+                }\n+            }\n             let inv_tau = 1.0 / tau;\n@@ -6137,6 +6151,38 @@ fn sae_decoder_lsq_init(\n                         a_init[[row, k]] = buf[k] / sum;\n                     }\n                 }\n+                if let Some(k_top) = top_k {\n+                    if k_top < k_atoms {\n+                        let mut paired: Vec<(f64, usize)> =\n+                            (0..k_atoms).map(|k| (a_init[[row, k]], k)).collect();\n+                        let cmp = |a: &(f64, usize), b: &(f64, usize)| {\n+                            b.0.partial_cmp(&a.0)\n+                                .unwrap_or(std::cmp::Ordering::Equal)\n+                                .then(a.1.cmp(&b.1))\n+                        };\n+         
</details>

_Codex-generated; pending adversarial audit before merge._

### CODE REVIEW

**Comment by SauersML** [comment]
## Fix is correct and root-cause; the regression test does not gate the defect

The core change is right and SPEC-clean: the seed/EM path was optimizing a *different* objective (dense softmax responsibilities) than the capped Arrow-Schur solve (`top_k` active set), and installing `set_softmax_active_cap(top_k)` before the cold routing refinement + projecting the seed LSQ responsibilities onto the top-k support closes that mismatch. Moving the cap earlier preserves the "after-fit projection is a no-op at the optimum" property (the cap persists), and threading `top_k` through `sae_decoder_lsq_init` / `sae_em_refine_routing_seed` / `sae_manifold_fit_minimal` is consistent. No magic constants, no grid search, no FD, no weakened existing tests. Good.

**But the new test is non-discriminating.** `sae_decoder_lsq_seed_honors_softmax_top_k_support_2132` uses logits `[8, -8]` with `tau = 1.0`, so the off-atom softmax mass is `e^-16 ≈ 1.1e-7`. Dense softmax is already ≈ hard top-1, so the near-orthogonal joint LSQ design is well-conditioned and yields `d0 ≈ 1`, `d1 ≈ -1` **with or without** the top-k projection. The assertions (`>0.99` / `<-0.99`) pass even if the projection were a no-op — it does not exercise the co-collapse the fix prevents.

To make it a real fails-before/passes-after test: use *moderate* logits (e.g. `~[0.5, -0.5]`) where dense softmax leaves both atoms materially active on shared rows and the uncapped joint seed genuinely co-collapses (near-identical `d0`, `d1`), then assert the `top_k=1` seed separates them. That would fail on base and pass here.

The fix itself is mergeable once the test actually gates the defect. Holding on the test only.

— panel review (independent seat)

**Comment by SauersML** [comment]
**Verdict: EDIT-THEN-MERGE.**

The fix is root-cause: the cold seed path (decoder LSQ + EM refinement) trained under dense softmax responsibilities while the Arrow–Schur solve applied a `top_k` active-set cap, so the seed optimized a different objective and could reintroduce the uniform co-collapse saddle. Threading `top_k` into `sae_decoder_lsq_init` / `sae_em_refine_routing_seed` and installing `set_softmax_active_cap(top_k)` before the cold refinement makes init see the same capped support. No forbidden patterns (no FD/autodiff in prod, no grid search, no wall-clock deadline, no magic-constant knob — the existing seed ridge is unchanged).

One blocker: the regression test used ±8 logits, where dense softmax ≈ hard top-1, so the seed LSQ landed on ±1 (coth(8)≈1.0000002) **even with the projection removed** — a no-op passed the one-sided `>0.99` asserts. Fixed on the branch: moderate ±0.5 logits (soft responsibilities), two-sided asserts around ±1, plus a `top_k=None` baseline arm asserting the dense fit stays coupled at coth(0.5)≈2.16. Now genuinely fails-before / passes-after and the projection is provably load-bearing on the fixture.

`Fixes #2132` → `Refs #2132`: the fix closes the cold-joint seed co-collapse, but the issue also tracks two lower-severity sub-findings (stagewise still below the linear-PCA floor; the OOS `reconstruct` shape-mismatch family) that this PR does not address.

— panel review, seat salvage
